primitives/src/vid/advz/payload_prover.rs

```diff
@@ -0,0 +1,505 @@
+// Copyright (c) 2023 Espresso Systems (espressosys.com)
+// This file is part of the Jellyfish library.
+
+// You should have received a copy of the MIT License
+// along with the Jellyfish library. If not, see <https://mit-license.org/>.
+
+//! Implementations of [`PayloadProver`] for `Advz`.
+//!
+//! Two implementations:
+//! 1. `PROOF = `[`Proof`]: KZG batch proof for the data. Useful for small
+//!    sub-slices of `payload` such as an individual transaction within a block.
+//!    Not snark-friendly because it requires a pairing.
+//! 2. `PROOF = `[`CommitRecovery`]: Rebuild the KZG commitment. Useful for
+//!    larger sub-slices of `payload` such as a complete namespace.
+//!    Snark-friendly because it does not require a pairing.
+
+use ark_poly::EvaluationDomain;
+use jf_utils::{bytes_to_field, compile_time_checks};
+
+use super::{
+    AffineRepr, Debug, DenseUVPolynomial, Digest, DynDigest, GenericAdvz, MerkleTreeScheme,
+    PolynomialCommitmentScheme, PrimeField, UnivariatePCS, Vec, VidResult, Write,
+};
+use crate::{
+    alloc::string::ToString,
+    vid::{payload_prover::PayloadProver, vid, VidError, VidScheme},
+};
+use ark_std::{format, ops::Range};
+
+impl<P, T, H, V> PayloadProver<Proof<P::Proof>> for GenericAdvz<P, T, H, V>
```

Comment:
Something slightly annoying about this interface is as a downstream user I'm probably going to have to write out whatever `P::Proof` is in order to help the typechecker figure out which kind of proof I want, since there are two different implementations with the same interface. In practice I'm probably going to be referring to the VID scheme itself through a type alias with specific instantiations of `<P, T, H, V>`. Makes it kind of annoying to click through rust docs to try and figure out what `P` is and then what `P::Proof` is so I can name this `Proof<P::Proof>` type.

Two ideas for making this a bit more ergonomic:

1. Bring back the monolothic prover trait with two pairs of methods with identical interfaces, but different names. That way it's obvious from the name of the method which type of proof I'm asking for, I don't have to actually write down the type. However, this would enshrine in the trait interface the idea that we have these two separate types of proofs, which might not be true for every conceivable VID scheme.
2. Introduce a marker type which we use to denote the kind of proof we're looking for abstractly, decoupled from the specific type we use to implement that proof. The marker type can be used to select the trait implementation I want and the actual type of proof can be an associated type of that trait. Something roughly like this:

```rust
trait PayloadProver<ProofType>: VidScheme {
  type Proof;
  fn payload_proof(...) -> VidResult<Self::Proof>;
}

struct SnarkFriendlyProof;
struct GranularProof;

impl PayloadProver<GranularProof> for GenericAdvz<P, T, H, V> {
  type Proof = Proof<P::Proof>;
  ...
}

impl PayloadProver<GranularProof> for GenericAdvz<P, T, H, V> {
  type Proof = CommitRecovery<P::Evaluation>;
  ...
}
```

Use like
```rust
PayloadProver::<GranularProof>::payload_proof(&scheme, payload, range);
PayloadProver::<SnarkFriendlyProof>::payload_proof(&scheme, payload, range);
```

Reply:
I like the marker type idea and I think we should use it if this is indeed a problem. But are you sure it will be so bad for downstream users? My tests work with the following:
```rust
let data_proof: Proof<_> = advz.payload_proof(payload, range)?;
let chunk_proof: CommitRecovery<_> = advz.payload_proof(payload, range)?;
```
Admittedly the names `Proof` and `CommitRecovery` kinda suck. It's hard to come up with good names without mentioning "namespace" or "transaction". 😉

Reply:
Ah the `<_>` is a nice trick. I think that's fine then